### DISCUSSION TITLE
How to use LuaJIT with sol2

### DISCUSSION THREAD

**Discussion by pfirsich** [microsoft/vcpkg#31702] [General]
PR #5863, motivated by #4019 moved LuaJIT headers to its own directory, i.e. `luajit/lua.h`, but sol2 searches for Lua and LuaJIT headers at `lua.h`. How do I use these together now? I think the headers should never have been moved as LuaJIT is a drop-in *replacement* for Lua, so it makes sense you can't install them both at the same time.

**Comment by dg0yt** [comment]
What exactly do you want to achieve - build port sol2 with luajit instead of lua?

vcpkg can't deal well with alternatives. There is a number of obstacles.
- If two ports can't be installed at the same time, they can't be both in CI.
- Port sol2 must use a single implementation of lua as dependency.
- luajit has restrictions with regard to cross builds, so it is not a drop-in replacement for all use cases.

Maybe it is possible to create an overlay port for the `lua` name which depends on `luajit` and simply copies files as needed. But I can't give an example.

**Reply by pfirsich** [reply]
sol2 is a library to aid in binding Lua code to C++ code. It officially support LuaJIT and I would like to use sol2 to bind to Lua, but use LuaJIT instead. I don't want to install both Lua and LuaJIT, just LuaJIT, but I want it to work with sol2. Simply installing both sol2 and luajit with vcpkg does not give the desired result, because sol2 attemps to include LuaJIT exactly like lua, by including `<lua.h>`. Is it clear now what I want to do?

**Reply by pfirsich** [reply]
I will add my response from the PR here as well, because the discussion has moved and I think it is worth adressing:
> sol2 depends on **a** Lua implementation. Officially supported are PUC Lua, LuaJIT and MoonJIT. I would say lua is not any more a dependency of sol2 than luajit is. sol2 just needs one (any) of them and I think it should be supported to use sol2 with any of them when installed through vcpkg.

**Reply by dg0yt** [reply]
This confirms my assumptions.
Port sol2 is written to use port lua, There is no supported way to switch to an alternative lua por except for switching via overlay ports.
This situation isn't limited to lua but also valid for qt, blas, jpeg etc.
Note that this isn't just about locating headers but also about properly tracking ABI and being able to compose matching binary artifacts from cache.

**Reply by pfirsich** [reply]
So do I understand correctly that using sol2 with LuaJIT is not supported by vcpkg at this moment, except via "overlay ports"? How do I use them?

**Comment by dg0yt** [comment]
Wait, sol2 is header-only. The lua dependency is declared only for Windows. 
So everything is assembled in the user project?
Did you try to add the luajit include directory at the beginning of the include search path?

**Reply by pfirsich** [reply]
I am using meson and simply point it to the .pc files. Do you suggest that adding an include path manually is the recommended solution?

**Reply by dg0yt** [reply]
The luajit .pc file should be okay. (Check it.) The key is to have the luajit include dir before any other vcpkg include dir, if you want luajit headers before lua headers. Probably this can be achieved with pc files in meson, too.

**Comment by yhyu13** [comment]
Here is my vcpkg.json file that works
```
{
    "dependencies": [
        {
            "name": "luajit",
            "features": [
                "buildvm-64"
            ],
            "default-features": true
        },
        "sol2"
    ],
    "builtin-baseline": "53bef8994c541b6561884a8395ea35715ece75db"
}
```

And here the cmake/findluajit.cmake file to support find_package of luajit https://github.com/yhyu13/HLVM-Engine/blob/d9e08c9b6fe6a3917f22626ce9e068457ff1c851/Engine/Source/Common/cmake/Findluajit.cmake

And all you need to do is include 
```
#include <sol/sol.hpp>
```

1, **But one caveat thing is that luajit is not automatically enabled when used by sol2 (even compiled with laujit)**

You need to open ffi and jit to enable luajit with sol2
```
lua.open_libraries(sol::lib::base, sol::lib::ffi, sol::lib::jit);
```

2, **Another issue is that you must link with dynamic lib in order to load C Modules**
vcpkg build luajit in dll on windows by default, but .a on linux, so we have to modify the port file for luajit, see [here](https://github.com/microsoft/vcpkg/issues/21917#issuecomment-2105702823)
